### PR TITLE
pkg/kubelet/server: ensure tests compatible with net/http.ServeMux

### LINKED ISSUE #46282: tests: please update pkg/kubelet/server TestServeExecInContainer and TestServeAttachContainer for Go1.9
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):
TestServeExecInContainer
response location

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
* BUG REPORT

**Kubernetes version** (use `kubectl version`):
Kubernetes at commit 8bee44b65f2fbf170f70c535b5ec1b3cbfad44b6
Sorry, I haven't installed kubectl, I was just wandering around the code

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
OS X
- **Kernel** (e.g. `uname -a`):
Darwin Emmanuels-MacBook-Pro-2.local 15.6.0 Darwin Kernel Version 15.6.0: Mon Aug 29 20:21:34 PDT 2016; root:xnu-3248.60.11~1/RELEASE_X86_64 x86_64
- **Install tools**:
Installed from source
- **Others**:

**What happened**:
Tests:
*  TestServeExecInContainer
*  TestServeAttachContainer
fail as of Go tip https://github.com/golang/go/commit/6a6c792eef55eded7fb3165a330ec2b239b83960 which will be part of Go1.9

**What you expected to happen**:
Those tests to pass

**How to reproduce it** (as minimally and precisely as possible):
Run tests:
*  TestServeExecInContainer
*  TestServeAttachContainer

**Anything else we need to know**:
The tests seem to be adapted to an old/buggy behavior of Go's http.ServeMux not forwarding along
the query string during redirects as per https://github.com/golang/go/issues/17841.
The bug in Go has since been fixed by CL https://golang.org/cl/43779 and I got this failure when running `make test` in Kubernetes
in particular
```
server_test.go:1327: 7: response location: expected http://localhost:12345/exec, got http://localhost:12345/exec?ignore=1&command=ls&command=-a&output=1
```
or in detail
```shell
E0523 04:03:05.187580   89629 server.go:244] Authorization error (user=test, verb=, resource=, subresource=)%!(EXTRA *errors.errorString=Failed)
E0523 04:03:05.318837   89629 server.go:656] you must specify at least 1 of stdin, stdout, stderr
--- FAIL: TestServeExecInContainer (0.03s)
	server_test.go:1327: 7: response location: expected http://localhost:12345/exec, got http://localhost:12345/exec?ignore=1&command=ls&command=-a&output=1
E0523 04:03:05.351045   89629 server.go:618] you must specify at least 1 of stdin, stdout, stderr
--- FAIL: TestServeAttachContainer (0.03s)
	server_test.go:1327: 7: response location: expected http://localhost:12345/attach, got http://localhost:12345/attach?ignore=1&output=1
E0523 04:03:05.528501   89629 server.go:730] query parameter "port" cannot be empty
E0523 04:03:05.530809   89629 server.go:730] unable to parse "abc" as a port: strconv.ParseUint: parsing "abc": invalid syntax
E0523 04:03:05.532914   89629 server.go:730] unable to parse "-1" as a port: strconv.ParseUint: parsing "-1": invalid syntax
E0523 04:03:05.534807   89629 server.go:730] unable to parse "65536" as a port: strconv.ParseUint: parsing "65536": value out of range
E0523 04:03:05.536912   89629 server.go:730] port "0" must be > 0
```
or
screen shot 2017-05-23 at 4 14 28 am

Prior to that CL, the test passes
```shell
ok  	k8s.io/kubernetes/pkg/kubelet/server	0.525s
```

I am also going to add a small memo to the Go project's bug tracker that that fix broke things here, just to keep vigilant.

### PR DESCRIPTION
**Pull request by odeke-em**
Fixes #46282.

After https://github.com/golang/go/issues/17841 was fixed as part
of Go1.9, Go's http.ServeMux now forwards along the query string
during redirects. The tests pre-Go1.9 relied on the old/buggy
behavior.

This PR ensures both forward and backward compability so that
if the new redirect Location is received with either pre-Go1.9 behavior
or Go1.9+ behavior, we can still properly test for both.

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by odeke-em** [comment]
CLA signed, please recheck.

**Comment by cmluciano** [comment]
@kubernetes/sig-node-pr-reviews

**Comment by cmluciano** [comment]
@k8s-bot ok to test

**Comment by resouer** [comment]
@k8s-bot ok to test

File: pkg/kubelet/server/server_test.go [hunk 1]
```diff
@@ -1143,6 +1143,26 @@ func TestServeExecInContainerIdleTimeout(t *testing.T) {
...
+	// new behaviors are both accounted for in tests.
+	if !strings.HasPrefix(redirected, original) {
+		return false
+	}
+	rest := strings.TrimPrefix(redirected, original)
+	// The remaining cases can be:
+	// 1. original and redirected are equal so no string left i.e ''
+	// 2. '/' is the leading character in the prefix i.e `/?foo=bar` or `/`
+	// 3. '?' is the leading character in the prefix i.e `?foo=bar`
+	return rest == "" || rest[0] == '/' || rest[0] == '?'
```
Comment (resouer) [inline pkg/kubelet/server/server_test.go]: This generally looks good, @timstclair mind to take a look?

**Comment by ncdc** [comment]
@odeke-em I don't think we need to try to preserve pre-1.9 backwards compatibility in the test case. This particular area of functionality always includes the query parameters in the `Location` header if there is a redirect, and it does not rely on the pre-1.9 broken behavior to my knowledge (otherwise `kubectl exec` and `kubectl attach` wouldn't work).

Would you be able to modify the test so that the `fw.fakeKubelet.redirectURL` is configured based on the query parameters of `url`? You could change `tests.responseLocation` to `responseLocationPrefix` and then dynamically append `verb` and the url query parameters.

WDYT?

File: pkg/kubelet/server/server_test.go [hunk 2]
```diff
@@ -1143,6 +1143,26 @@ func TestServeExecInContainerIdleTimeout(t *testing.T) {
...
+	// expecting the old behavior will fail the tests.
+	// This function helps ensure that the old and
+	// new behaviors are both accounted for in tests.
+	if !strings.HasPrefix(redirected, original) {
+		return false
+	}
+	rest := strings.TrimPrefix(redirected, original)
+	// The remaining cases can be:
+	// 1. original and redirected are equal so no string left i.e ''
+	// 2. '/' is the leading character in the prefix i.e `/?foo=bar` or `/`
```
Comment (timstclair) [inline pkg/kubelet/server/server_test.go]: s/prefix/remainder/

same below

File: pkg/kubelet/server/server_test.go (hunk 1 above)
Comment (timstclair) [inline pkg/kubelet/server/server_test.go]: I'm not sure this is secure, but I think it's good enough for test code.

**Comment by odeke-em** [comment]
Thanks y'all for the first round of review.

@ncdc in regards to
```
@odeke-em I don't think we need to try to preserve pre-1.9 backwards compatibility in the test case. This particular area of functionality always includes the query parameters in the Location header if there is a redirect, and it does not rely on the pre-1.9 broken behavior to my knowledge
```
Hmm, so if you run that test with preGo1.9, it fails with the error that I reported previously because it didn't expect the query string to be forwarded along. My reasoning for accounting for the preGo1.9 behavior is that not everyone hacking on Kubernetes will switch to Go1.9 when it is released and also to enable those using Go tip to still be able to hack on Kubernetes. 

```
 (otherwise kubectl exec and kubectl attach wouldn't work).
```
I didn't run kubectl exec nor kubectl attach so I can't judge if those will fail.

```
Would you be able to modify the test so that the fw.fakeKubelet.redirectURL is configured based on the query parameters of url? You could change tests.responseLocation to responseLocationPrefix and then dynamically append verb and the url query parameters.

WDYT?
```
Unfortunately I am not that familiar around the code base so that might be too intricate for me. However, if you'd like, you can takeover the PR, I'd be happy with that.

@timothysc am going to update it in a second.

File: pkg/kubelet/server/server_test.go (hunk 2 above)
Comment (odeke-em) [inline pkg/kubelet/server/server_test.go]: Done, thanks.

**Comment by ncdc** [comment]
@odeke-em 

> Hmm, so if you run that test with preGo1.9, it fails with the error that I reported previously because it didn't expect the query string to be forwarded along. My reasoning for accounting for the preGo1.9 behavior is that not everyone hacking on Kubernetes will switch to Go1.9 when it is released and also to enable those using Go tip to still be able to hack on Kubernetes.

What I am saying is the _test_ is invalid. The redirect location in question (`localhost:12345/exec` or `localhost:12345/attach`) is a made-up bogus URL that only exists in this test suite.

I can put together an update later tonight or tomorrow to show you what I would do, if that works?

**Comment by odeke-em** [comment]
@ncdc despite being a fake URL, the unit test behavior was failing and without running kubectl, I unfortunately can't judge.
Aye aye, that works for me, thanks.

**Comment by odeke-em** [comment]
@ncdc gotcha 🍻😆, makes sense.

**Comment by odeke-em** [comment]
Btw no rush y'all we've reverted the fix due to some unfortunate technical difficulties,
plus we've got tonnes of time before Go1.9 comes out :)

**Comment by odeke-em** [comment]
Closing this issue since it won't be changed until Go1.10. We can resurrect it then! Thank you everyone for the reviews.

**Pull request closed without merging**